### PR TITLE
Issue 18686 apps select fields

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
We're adding support for apps to use select like fields. See https://github.com/dotCMS/core/issues/18686

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/security/apps/AbstractProperty.java [hunk 1]
```diff
@@ -1,8 +1,17 @@
 package com.dotcms.security.apps;
 
+import static com.liferay.util.StringPool.COLON;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AbstractProperty.java:3]: Codacy found an issue: [Avoid unused imports such as 'com.liferay.util.StringPool.COLON'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/security/apps/AbstractProperty.java [hunk 2]
```diff
@@ -1,8 +1,17 @@
...
+import static com.liferay.util.StringPool.COMMA;
+
 import com.dotmarketing.util.UtilMethods;
 import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.google.common.collect.ImmutableMap;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AbstractProperty.java:8]: Codacy found an issue: [Avoid unused imports such as 'com.google.common.collect.ImmutableMap'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/security/apps/ParamDescriptor.java [hunk 3]
```diff
@@ -3,13 +3,14 @@
...
 import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.List;
 
 /**
  * This is an implementation of a Parameter
  * Class used to collect properties or secrets set from the front-end
  * This is mostly used to pass values from the front-end into the Resource.
  */
-public class ParamDescriptor extends AbstractProperty<String>{
+public class ParamDescriptor extends AbstractProperty<Object>{
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/ParamDescriptor.java:13]: Codacy found an issue: [A class which only has private constructors should be final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java [hunk 4]
```diff
@@ -753,38 +767,89 @@ private boolean validateServiceDescriptor(final AppSchema appDescriptor)
...
+            }
         }
 
-        if (StringPool.NULL.equalsIgnoreCase(descriptor.getValue()) && descriptor.isRequired()) {
-            errors.add(String.format(
-                    "Null isn't allowed as the default value on required params see `%s`. ",
-                    name)
-            );
+        if(Type.STRING.equals(descriptor.getType())){
+            if(!(descriptor.getValue() instanceof String)){
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java:812]: Codacy found an issue: [These nested if statements could be combined](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java [hunk 5]
```diff
@@ -753,38 +767,89 @@ private boolean validateServiceDescriptor(final AppSchema appDescriptor)
...
+                        final Map map = (Map) object;
+                        if (!map.containsKey("label") || !map.containsKey("value") ) {
+                            errors.add(String.format("Malformed list. Param: `%s`. Every entry of the `List` has to have the following fields (`label`,`value`). ", name));
+                        }
+                         if(map.containsKey("selected")){
+                             selectedCount++;
+                         }
+                    }
+                }
+                if(selectedCount > 1 ){
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java:849]: Codacy found an issue: [Avoid using Literals in Conditional Statements](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/security/apps/AbstractProperty.java [hunk 6]
```diff
@@ -1,8 +1,17 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
 import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AbstractProperty.java:14]: Codacy found an issue: [Avoid unused imports such as 'java.util.stream.Stream'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java [hunk 7]
```diff
@@ -640,17 +641,28 @@ private static String getServiceDescriptorDirectory() {
...
-        try (final DirectoryStream<Path> stream = Files.newDirectoryStream(Paths.get(dir))) {
-            for (final Path path : stream) {
-                if (!Files.isDirectory(path)) {
-                    final String fileName = path.getFileName().toString();
-                    if (fileName.endsWith("yaml") || fileName.endsWith("yml")) {
-                        fileList.add(path.toString());
-                    }
-                }
-            }
+        try (final DirectoryStream<Path> stream = Files.newDirectoryStream(Paths.get(dir), filter)) {
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java:662]: Codacy found an issue: [Unnecessary modifier 'final' on resource specification 'stream': resource specifications are implicitly final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/security/apps/AbstractProperty.java [hunk 8]
```diff
@@ -1,8 +1,17 @@
 package com.dotcms.security.apps;
 
+import static com.liferay.util.StringPool.COLON;
+import static com.liferay.util.StringPool.COMMA;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AbstractProperty.java:4]: Codacy found an issue: [Avoid unused imports such as 'com.liferay.util.StringPool.COMMA'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java [hunk 9]
```diff
@@ -773,6 +773,92 @@ public void Test_Secret_Serializer_Returned_Values_Match_Descriptor_Verify_Hidde
...
+     * Given scenario: We create a secret then we update the select value sending one of the values
+     * Expected Result: Only the public value  was supposed to be updated. The other 2 hidden fields should remain the same since we sent a bunch of *****
+     * @throws IOException
+     */
+    @Test
+    public void Test_Select_List_Param()
+            throws IOException {
+       //Select list definition
+       final List<Map<String,String>> list = ImmutableList.of(
+           ImmutableMap.of("label","-","value",""),
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java:787]: Codacy found an issue: [The String literal "label" appears 4 times in this file; the first occurrence is on line 787](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/apps/view/SecretView.java [hunk 10]
```diff
@@ -1,5 +1,7 @@
 package com.dotcms.rest.api.v1.apps.view;
 
+import static com.liferay.util.StringPool.*;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/apps/view/SecretView.java:3]: Codacy found an issue: [Avoid unused imports such as 'com.liferay.util.StringPool'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/apps/view/SecretView.java [hunk 11]
```diff
@@ -12,11 +14,13 @@
 import com.fasterxml.jackson.databind.SerializerProvider;
 import com.fasterxml.jackson.databind.annotation.JsonSerialize;
 import com.google.common.annotations.VisibleForTesting;
+import com.liferay.util.StringPool;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/apps/view/SecretView.java:17]: Codacy found an issue: [Avoid unused imports such as 'com.liferay.util.StringPool'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/security/apps/ParamDescriptor.java [hunk 12]
```diff
@@ -3,13 +3,14 @@
 import com.dotmarketing.util.UtilMethods;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.List;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/ParamDescriptor.java:6]: Codacy found an issue: [Avoid unused imports such as 'java.util.List'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java [hunk 13]
```diff
@@ -773,6 +773,92 @@ public void Test_Secret_Serializer_Returned_Values_Match_Descriptor_Verify_Hidde
...
+           ImmutableMap.of("label","uno","value","1"),
+           ImmutableMap.of("label","dos","value","2","selected","true"),
+           ImmutableMap.of("label","tres","value","3")
+       );
+
+        final AppDescriptorDataGen dataGen = new AppDescriptorDataGen()
+                .stringParam("param1", false,  true)
+                .selectParam("selectParam", true, list)
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java:795]: Codacy found an issue: [The String literal "selectParam" appears 4 times in this file; the first occurrence is on line 795](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/security/apps/AbstractProperty.java [hunk 14]
```diff
@@ -1,8 +1,17 @@
...
+
 import com.dotmarketing.util.UtilMethods;
 import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.google.common.collect.ImmutableMap;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
 import java.util.Objects;
+import java.util.stream.Collectors;
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AbstractProperty.java:13]: Codacy found an issue: [Avoid unused imports such as 'java.util.stream.Collectors'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java (hunk 13 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java:795]: Codacy found an issue: [The String literal "selectParam" appears 4 times in this file; the first occurrence is on line 795](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java (hunk 9 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java:787]: Codacy found an issue: [The String literal "label" appears 4 times in this file; the first occurrence is on line 787](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

**Review by jdotcms — APPROVED** [review]

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java (hunk 13 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java:795]: Codacy found an issue: [The String literal "selectParam" appears 4 times in this file; the first occurrence is on line 795](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java (hunk 9 above)
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotcms/rest/api/v1/apps/AppsResourceTest.java:787]: Codacy found an issue: [The String literal "label" appears 4 times in this file; the first occurrence is on line 787](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java [hunk 15]
```diff
@@ -610,6 +611,45 @@ private void removeDescriptor(final AppDescriptor descriptor) throws DotDataExce
...
+     * Condition check This verifies the descriptor demands the param to be required but.. No default value is provided and the secret neither has a stored value
+     * @param descriptor ParamDescriptor
+     * @param secret stored secret
+     * @return
+     */
+    private boolean isRequiredWithNoDefaultValue(final ParamDescriptor descriptor, final Secret secret ){
+        //Verify we have a param marked required and no default Value
+        final boolean isRequiredWithNoDefaultParam = (descriptor.isRequired() && isEmpty(descriptor.getValue()));
+        //Verify the secret is empty
+        final boolean isSecretWithEmptyValue = (null == secret || isNotSet(secret.getValue()));
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java:624]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

File: dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java [hunk 16]
```diff
@@ -610,6 +611,45 @@ private void removeDescriptor(final AppDescriptor descriptor) throws DotDataExce
...
 
+    /**
+     * Condition check This verifies the descriptor demands the param to be required but.. No default value is provided and the secret neither has a stored value
+     * @param descriptor ParamDescriptor
+     * @param secret stored secret
+     * @return
+     */
+    private boolean isRequiredWithNoDefaultValue(final ParamDescriptor descriptor, final Secret secret ){
+        //Verify we have a param marked required and no default Value
+        final boolean isRequiredWithNoDefaultParam = (descriptor.isRequired() && isEmpty(descriptor.getValue()));
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotcms/security/apps/AppsAPIImpl.java:622]: Codacy found an issue: [Useless parentheses.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5895122)

**Pull request merged**